Emit a suggestion to cast `!` to a concrete type when `!` is infered as the hidden type for `impl Trait` and `!` doesn't implement `Trait`

### Code

```Rust
use std::ops::Add;

fn foo() -> impl Add<u32> {
    unimplemented!()
}

fn main() {}
```

### Current output

```text
error[E0277]: cannot add `u32` to `!`
 --> src/main.rs:3:13
  |
3 | fn foo() -> impl Add<u32> {
  |             ^^^^^^^^^^^^^ no implementation for `! + u32`
4 |     unimplemented!()
  |     ---------------- return type was inferred to be `!` here
  |
  = help: the trait `Add<u32>` is not implemented for `!`
```

### Desired output

A hint or suggestion should be added which explains that casting to a a concrete type which implements the given trait (for example, `unimplemented!() as u32`) will resolve the error, given that `!` can be coerced into any concrete type.


### Rationale and extra context

While the `unimplemented!` and `todo!` macros are usually temporary in code, using them to mock APIs is a common use case, such as allowing tests to typecheck or to write an API shape before writing the implementation.

When such a mocked function returns `impl Trait`, the compiler error makes it seem like it's not possible to leave these functions unimplemented. However, casting to a concrete type which implements `Trait` is a simple workaround that is also non-obvious.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: x86_64-unknown-linux-gnu
release: 1.96.0
LLVM version: 22.1.2
```

### Anything else?

Related to #36375 and #113875

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

This is strange, I'm getting a different output from the Rust compiler installed on my system. The given code, version and host are exactly the same.
```rs
error[E0277]: cannot add `u32` to `()`
 --> test.rs:3:13
  |
3 | fn foo() -> impl Add<u32> {
  |             ^^^^^^^^^^^^^ no implementation for `() + u32`
  |
  = help: the trait `Add<u32>` is not implemented for `()`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0277`.
```
```
rustc 1.96.0 (ac68faa20 2026-05-25)
binary: rustc
commit-hash: ac68faa20c58cbccd01ee7208bf3b6e93a7d7f96
commit-date: 2026-05-25
host: x86_64-unknown-linux-gnu
release: 1.96.0
LLVM version: 22.1.2
```
Can you check again on your end?

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs
@@ -4479,7 +4479,6 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }
             ObligationCauseCode::OpaqueReturnType(expr_info) => {
                 let (expr_ty, expr) = if let Some((expr_ty, hir_id)) = expr_info {
-                    let expr_ty = tcx.short_string(expr_ty, err.long_ty_path());
                     let expr = tcx.hir_expect_expr(hir_id);
                     (expr_ty, expr)
                 } else if let Some(body_id) = tcx.hir_node_by_def_id(body_id).body_id()
@@ -4494,15 +4493,27 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                     && let ty::ClauseKind::Trait(pred) = pred.kind().skip_binder()
                     && self.can_eq(param_env, pred.self_ty(), expr_ty)
                 {
-                    let expr_ty = tcx.short_string(expr_ty, err.long_ty_path());
                     (expr_ty, expr)
                 } else {
                     return;
                 };
+                let expr_ty_string = tcx.short_string(expr_ty, err.long_ty_path());
+                if expr_ty.is_never()
+                    && let span = expr.span.source_callsite()
+                    && let Ok(snippet) = tcx.sess.source_map().span_to_snippet(span)
+                    && span != expr.span
+                {
+                    err.span_suggestion(
+                        span,
+                        "`!` can be coerced to any type; consider casting it to a concrete type that implements the trait",
+                        format!("{snippet} as /* Type */"),
+                        Applicability::HasPlaceholders,
+                    );
+                }
                 err.span_label(
                     expr.span,
                     with_forced_trimmed_paths!(format!(
-                        "return type was inferred to be `{expr_ty}` here",
+                        "return type was inferred to be `{expr_ty_string}` here",
                     )),
                 );
                 suggest_remove_deref(err, &expr);

--- a/tests/ui/impl-trait/return-never-type.rs
+++ b/tests/ui/impl-trait/return-never-type.rs
@@ -1,0 +1,14 @@
+//@ edition:2024
+
+#![feature(never_type)]
+
+use std::ops::Add;
+
+fn foo() -> impl Add<u32> {
+    //~^ ERROR cannot add `u32` to `!`
+    //~| HELP the trait `Add<u32>` is not implemented for `!`
+    unimplemented!()
+    //~^ HELP `!` can be coerced to any type; consider casting it to a concrete type that implements the trait
+}
+
+fn main() {}

--- a/tests/ui/impl-trait/return-never-type.stderr
+++ b/tests/ui/impl-trait/return-never-type.stderr
@@ -1,0 +1,18 @@
+error[E0277]: cannot add `u32` to `!`
+  --> $DIR/return-never-type.rs:7:13
+   |
+LL | fn foo() -> impl Add<u32> {
+   |             ^^^^^^^^^^^^^ no implementation for `! + u32`
+...
+LL |     unimplemented!()
+   |     ---------------- return type was inferred to be `!` here
+   |
+   = help: the trait `Add<u32>` is not implemented for `!`
+help: `!` can be coerced to any type; consider casting it to a concrete type that implements the trait
+   |
+LL |     unimplemented!() as /* Type */
+   |                      +++++++++++++
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0277`.

